### DISCUSSION TITLE
Ensuring Consistent Internal Responses in OpenFEDEM

### DISCUSSION THREAD

**Discussion by oeysstra** [openfedem/fedem-solvers#8] [General]
In OpenFEDEM, the solver process involves several critical steps:

 - Super-Element Reduction of the Parts
 - Dynamic Solution of the System
 - Recovery of the Internal Deflections and Stresses
 
The default representation of super-elements is through Nastran BDFs. If these files also contain load specifications, reduced load vectors are created. These load vectors can be scaled with a time series function.

While the loads are part of the system, ensuring that the system response (i.e., the System DoFs) is accurate, there is a notable issue: the internal response, also known as the secondary solution, is missing. As a result, we observe an inconsistent deflection and stress representation for models that contain parts with loads specified in the Nastran BDF files.

**Discussion Points:**
How can we ensure that the internal response is accurately recovered in the presence of loads specified in Nastran BDF files?

**Comment by kmokstad** [comment]
There are two levels in the solution to this issue:

1. Calculation of local displacements at internal nodal points accounting for the superelement loads.
2. Calculation of pointwise deformations the stress evaluation points accounting for the secondary solution within the elements.

For the first point, we are nearly there. As of now, only the distributed gravitation loads are accounted for in the recovery of internal nodal displacements. This is done according to equation (3.5) in the [FEDEM Theory guide](https://github.com/openfedem/fedem-docs/releases/download/fedem-8.0.7/FedemTheoryGuide.pdf)  (see page 3-10). This will give the local displacements of all internal nodal points due to the gravity load, which will be a constant addition to the displacements due to the motion in the supernodes (Triads) of the FE part. This displacement vector is calculated by the FE part reducer and stored in an fmx-file, such that the recovery processes only need to read those files to account for them. This calculation is not performed when component modes are used (activated by specifying a negative number of component modes for the Part).

This can be extended to also cover any occurring reduced load vector. Each load case will then result in one such displacement vector (again computed by the FE part reducer). The recovery modules then only need to read in those displacement vectors, and apply the scaling according to the load history to obtain the true displacement of all internal nodes.

The second point is a bit more involved, as we then need to go into the finite element implementation itself to calculate the consistent displacements due to an occurring distributed load. These contributions will, however, be small if the FE mesh is sufficiently refined. Therefore, I don't think it is worthwhile going into this in general.
